### PR TITLE
staticcheck fixes and addition to CI

### LINKED ISSUE #56: StaticCheck test results (26 errors)
I've just been doing a little bit of research on various code testing and analysis tools for golang and ran the https://staticcheck.io tool against the Smart Gateway. The following is the output analysis. Will require further investigation to determine if these items are valid and should be addressed during a code refactor.

```
staticcheck -f stylish github.com/redhat-service-assurance/smart-gateway/...
/home/lmadsen/src/go/src/github.com/redhat-service-assurance/smart-gateway/internal/pkg/alerts/alerts.go
  (25, 5)  U1000  var alertStatus is unused

/home/lmadsen/src/go/src/github.com/redhat-service-assurance/smart-gateway/internal/pkg/api/handler.go
  (68, 6)  U1000  func healthzHandler is unused

/home/lmadsen/src/go/src/github.com/redhat-service-assurance/smart-gateway/internal/pkg/cacheutil/cacheserver.go
  (15, 5)    U1000  var debugc is unused
  (117, 2)   S1008  should use 'return <expr>' instead of 'if <expr> { return <bool> }; return <bool>'
  (175, 2)   U1000  field mincollectinterval is unused
  (211, 23)  U1000  func CacheServer.close is unused

/home/lmadsen/src/go/src/github.com/redhat-service-assurance/smart-gateway/internal/pkg/events/events.go
  (51, 2)    U1000   var amqpHandler is unused
  (53, 2)    U1000   var elasticClient is unused
  (182, 27)  SA1019  prometheus.Handler is deprecated: Please note the issues described in the doc comment of InstrumentHandler. You might want to consider using promhttp.Handler instead. 
  (193, 21)  S1025   the argument is already a string, there's no need to use fmt.Sprintf
  (254, 3)   SA5004  should not have an empty default case in a for+select loop. The loop will spin.

/home/lmadsen/src/go/src/github.com/redhat-service-assurance/smart-gateway/internal/pkg/metrics/metrics.go
  (26, 2)    U1000   var cacheServer is unused
  (30, 2)    U1000   var amqpHandler is unused
  (31, 2)    U1000   var myHandler is unused
  (32, 2)    U1000   var hostwaitgroup is unused
  (169, 5)   S1002   should omit comparison to bool constant, can be simplified to !serverConfig.CPUStats
  (176, 29)  SA1019  prometheus.Handler is deprecated: Please note the issues described in the doc comment of InstrumentHandler. You might want to consider using promhttp.Handler instead. 

/home/lmadsen/src/go/src/github.com/redhat-service-assurance/smart-gateway/internal/pkg/saelastic/client.go
  (45, 2)   U1000   field err is unused
  (138, 2)  SA9003  empty branch
  (199, 2)  SA9003  empty branch

/home/lmadsen/src/go/src/github.com/redhat-service-assurance/smart-gateway/internal/pkg/saelastic/utils.go
  (76, 6)   U1000  func typeSwitch is unused
  (80, 11)  S1034  assigning the result of this type assertion to a variable (switch val := val.(type)) could eliminate the following type assertions:
            /home/lmadsen/src/go/src/github.com/redhat-service-assurance/smart-gateway/internal/pkg/saelastic/utils.go:82:19

/home/lmadsen/src/go/src/github.com/redhat-service-assurance/smart-gateway/internal/pkg/tsdb/prometheus.go
  (29, 2)   S1021   should merge variable declaration with assignment on next line
  (53, 25)  ST1005  error strings should not be capitalized
  (65, 5)   S1002   should omit comparison to bool constant, can be simplified to usetimestamp

/home/lmadsen/src/go/src/github.com/redhat-service-assurance/smart-gateway/tests/internal_pkg/saconfig_test.go
  (72, 6)  U1000  type LoadFunc is unused

 ✖ 26 problems (26 errors, 0 warnings)
```

### PR DESCRIPTION
**Pull request by csibbitt**
```
[csibbitt@laptop smart-gateway (csibbitt-717-static-checks) ]$ staticcheck -f stylish ./...
-
  (0, 0)  compile  could not load export data: no export data for "github.com/redhat-service-assurance/smart-gateway/tests/internal_pkg"

 ✖ 2 problems (1 errors, 0 warnings)
```

Unit tests still pass.

I tested this branch in my perftest harness to make sure all the same metrics are available from the internal instrumentation and see if there is any difference in scrape times (I doubt it)

Closes #56

### CODE REVIEW

File: internal/pkg/cacheutil/cacheserver.go [hunk 1]
```diff
@@ -208,23 +203,16 @@ func (cs *CacheServer) Put(incomingData incoming.DataTypeInterface) {
...
 		// Reuse buffer if there's room.
 		buffer := <-cs.ch
 		shard := cs.cache.GetShard(buffer.data.GetKey())
 		shard.SetData(buffer.data)
 		select {
 		case freeList <- buffer:
-		// Buffer on free list; nothing more to do.
-		case <-quitCacheServerCh:
-			break LOOP
+			// Buffer on free list; nothing more to do.
```
Comment (csibbitt) [inline internal/pkg/cacheutil/cacheserver.go]: This is a bit weird to me, will need to look closer and what this does. Basically what we have here is an attempt to read from buffer into freeList, and if either there is no message in buffer, or no space in freelist, we just carry on. Naively, this seems like an odd thing to do, but I haven't looked any deeper.

The mechanism for exiting this main cacheserver loop was never implmented (nothing ever calls close()) and I don't really see why you'd want this to shut down during operation.

File: internal/pkg/cacheutil/cacheserver.go [hunk 2]
```diff
@@ -208,23 +203,16 @@ func (cs *CacheServer) Put(incomingData incoming.DataTypeInterface) {
...
 }
 
-func (cs CacheServer) close() {
-	quitCacheServerCh <- true
-	close(quitCacheServerCh)
-}
-
 func (cs CacheServer) loop() {
-LOOP:
+	debugc("Debug:CacheServer loop started")
```
Comment (csibbitt) [inline internal/pkg/cacheutil/cacheserver.go]: This is to eliminate the "debugc is unused" warning

**Comment by csibbitt** [comment]
CI is broken on dep packaging; I will fix.

Also, I moved to the recommended version of staticcheck (https://staticcheck.io/docs/#installation) and the error I was getting went away. Curiously it still says "1 problems" but I don't know what it is.

```
[csibbitt@laptop smart-gateway (csibbitt-717-static-checks) ]$ staticcheck --version
staticcheck 2019.2.2

[csibbitt@laptop smart-gateway (csibbitt-717-static-checks) ]$ staticcheck -f stylish -tests=false ./...
 ✖ 1 problems (0 errors, 0 warnings)
[csibbitt@laptop smart-gateway (csibbitt-717-static-checks) ]$ echo $?
0
```

File: pre-commit [hunk 3]
```diff
@@ -1,9 +1,10 @@
-#!/bin/sh
+#!/bin/bash
```
Comment (csibbitt) [inline pre-commit]: `[[` on L34 is not a POSIX shell feature according to shellcheck.

File: internal/pkg/events/events.go [hunk 4]
```diff
@@ -190,11 +189,11 @@ func StartEvents() {
 																	</body>
 																	</html>`))
 		})
-		APIEndpointURL := fmt.Sprintf("%s", serverConfig.API.APIEndpointURL)
-		go func(APIEndpointURL string) {
+		go func() {
+			APIEndpointURL := serverConfig.API.APIEndpointURL
```
Comment (csibbitt) [inline internal/pkg/events/events.go]: No need to pass this variable into the closure

**Comment by csibbitt** [comment]
Unfortunately the Prometheus related changes broke (a non-critical) part of my perftest dashboard. It seems that http_request_duration_microseconds, http_request_duration_microseconds_count, and http_request_duration_microseconds_sum are no longer being scraped from the SG.

I'm okay with merging this as-is anyways; the fix might be on the dashboard side anyways, and if any changes are required here I'll open another PR if this one is already merged.

File: internal/pkg/saelastic/utils.go [hunk 5]
```diff
@@ -72,24 +69,21 @@ func typeSwitchAlertname(tst interface{}) (string, IndexType, error) {
 	return string(GENERICINDEX), GENERICINDEXTYPE, nil
 }
 
-// Not used
+//lint:ignore U1000 we might be able to delete this, but we should dig into the history of what it's for
```
Comment (leifmadsen) [inline internal/pkg/saelastic/utils.go]: Well this function is definitely related to the collectd work for the connectivity, sysevent, and procevent plugins, but I have no idea why...

@aneeshkp ?

File: pre-commit (hunk 3 above)
Comment (leifmadsen) [reply pre-commit]: True story :)

**Comment by leifmadsen** [comment]
Feel free to merge this, but I wouldn't mind a review from Martin first. I expect we can merge this Friday.

**Comment by leifmadsen** [comment]
@csibbitt one last thing, if one of your commit messages ends in `Closes #56` then that issue will be automatically closed on merge to `master`.

**Comment by csibbitt** [comment]
I updated this to squash a conflict, and now the build is failing on an unrelated thing:

```
 *** Uploading coverage report to coveralls
+ goveralls -coverprofile=coverage.out
15:20:57 gitinfo.go:56: exit status 129: error: unknown option `format'
usage: git branch [options] [-r | -a] [--merged | --no-merged]
   or: git branch [options] [-l] [-f] <branchname> [<start-point>]
   or: git branch [options] [-r] (-d | -D) <branchname>...
   or: git branch [options] (-m | -M) [<oldbranch>] <newbranch>
```

File: internal/pkg/cacheutil/cacheserver.go (hunk 1 above)
Comment (paramite) [reply internal/pkg/cacheutil/cacheserver.go]: So I'm not that skilled in Go to give my hand to fire for the following senstence, but: CacheServer.loop is executed in separate goroutine and as such should be explicitly ended so we won't leave zombie threads after the smart gateway ends (either by accident or on purpose). So IMO the fact that the close method is not called is the actual bug :).

Comment (paramite) [reply internal/pkg/cacheutil/cacheserver.go]: That said we need to defer close instead.

Comment (csibbitt) [reply internal/pkg/cacheutil/cacheserver.go]: I'm not really worried about leaking threads, this loop runs for the lifetime of the process and when the parent ends, so do it's threads (our containers DO shut down), that said, I think you're totally right and a deferred close is proper. I was not attempting to alter or fix any functionality with this PR, just address the staticcheck errors, but this seems small enough that I could add it.

Comment (csibbitt) [reply internal/pkg/cacheutil/cacheserver.go]: Is this really a concern? NewCacheServer() is called just once from StartMetrics(), and it spawns the goroutine that runs this infinite loop. When StartMetrics() returns, main() will exit (https://github.com/redhat-service-assurance/smart-gateway/blob/master/cmd/main.go#L19). When the main goroutine exits, so do all other goroutines (https://golang.org/ref/spec#Program_execution) this happens abruptly without calling defers in the other goroutines. They do not leak.

Zombies processes are when a child exits and the parent does not wait on it to consume (or ignore) it's exit code; in terms of a pattern, that's actually the opposite of what's happening here! Though it's worth mentioning that there are no "parent" and "child" goroutines anyways, and goroutines are not processes (or threads), so they do not follow the unix process pattern of needing a wait() or equivalent SIGCHLD handling.

Functionally speaking, the smart gateway container exits when we tell it to, so I don't think there is a bug.

Our containers DO shut down when we tell them to.

Comment (paramite) [reply internal/pkg/cacheutil/cacheserver.go]: Goroutines runtimes are scheduled on OS threads IIUC, but you're right ... still it does not mean that it is not an antipattern :) [1][2], 

[1] https://dave.cheney.net/2016/12/22/never-start-a-goroutine-without-knowing-how-it-will-stop
[2] https://stackoverflow.com/questions/25518531/ok-to-exit-program-with-active-goroutine

**Comment by paramite** [comment]
Sorry, github reviews UI is pain for me.

**Comment by leifmadsen** [comment]
@csibbitt is this anything we should worry about working towards a release for? I assume this is just code clean up and we can deal with it whenever the next release is justified for functionality/fixes?

**Comment by csibbitt** [comment]
Definitely nothing we need to worry about releasing in here; functionality was unchanged, just trying to get good CI checks in place.

**Pull request merged**